### ISSUE TITLE
add SQL support for unsigned integers

### ISSUE THREAD

**Issue by kmitchener** [apache/arrow-datafusion#3325]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

Related to PR #3309 which wants to add SQL support for tinyint and unsigned ints. The question is whether we want to support SQL access to underlying data types that Arrow supports. 

Here's a table of integer data type support between Spark, Postgres, DataFusion, Arrow, and Parquet:
![image](https://user-images.githubusercontent.com/692497/187908641-8ed64242-a7a0-4f78-8b10-ef1ecadf6053.png)

"SQL support" in this case means only 3 things, as far as I can tell:
* support for the data type in the `CREATE EXTERNAL TABLE x (schema..)` syntax
* support for the data type in the SQL `CAST` function 
* no memory table support via `CREATE TABLE x AS SELECT 1::tinyint unsigned` syntax

Otherwise, the unsigned variants are already supported by DataFusion, because Arrow supports them. Unsigned integers can be in custom table providers, or in a parquet file. 

Without SQL support for these data types, we won't be able to cast to unsigned types from SQL, nor load them from CSV/JSON.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by kmitchener** [comment]
@liukun4515 @alamb for discussion

**Comment by alamb** [comment]
Thank you for this wonderful writeup @kmitchener 

I personally don't have a strong opinion one way or the other in terms of how important it is to support unsigned variants. My default position is if someone finds `unsigned` support useful then we should support it

The most realistic usecase I can imagine is creating `RecordBatch`es that are passed to some other system that expect `unsigned` types for some reason

**Comment by kmitchener** [comment]
It seems to me to be a useful extension, especially for testing and the cast() support. I doubt there are real-world situations where someone is concerned about memory usage of ints but is reading data from a CSV or JSON. On the other hand, I don't see the harm in supporting a superset of functionality that directly supports the Arrow types DataFusion is based on.

**Comment by liukun4515** [comment]
> It seems to me to be a useful extension, especially for testing and the cast() support. I doubt there are real-world situations where someone is concerned about memory usage of ints but is reading data from a CSV or JSON. On the other hand, I don't see the harm in supporting a superset of functionality that directly supports the Arrow types DataFusion is based on.

@kmitchener Do you have some case using the unsigned value?

@alamb 
As I know, the PG don't support the unsigned value in the SQL syntax.
I think we should mainly rely on one sql dialect or sql syntax.

**Comment by liukun4515** [comment]
CC @andygrove

**Comment by liukun4515** [comment]
@alamb  we can bring this discussion to the slack channel

**Comment by kmitchener** [comment]
@liukun4515 Yes, I'd like to be able to cast() from SQL to these datatypes that are supported by DataFusion functions. Primarily for testing or evaluating behavior or for examples in documentation. Or to load directly from CSV into the unsigned form that Arrow supports (again for testing functions that support those types).

Seems DataFusion is used a lot for parquet files -- wouldn't it make sense that if you can select sum(unsigned int), that you could also cast to it?

I fully understand and support that we are trying to mirror PG syntax. But I don't think that should prevent us from supporting a superset of data types that PG doesn't. Especially when the support already exists in Arrow and DataFusion functions and we just need this thin shim to extend the support to SQL.

**Comment by kmitchener** [comment]
Maybe we should also extend this discussion to the other data types? let me continue to build out that spreadsheet for time / string / binary, etc and then we can have a fuller discussion of all the datatypes and SQL support, etc. For example, I know support for interval casting and time is currently missing and desired.

**Comment by kmitchener** [comment]
https://docs.google.com/spreadsheets/d/1jufe8vtmePxcgUdml_IFJS648mluZVUc/edit?usp=sharing&ouid=114751397656458197444&rtpof=true&sd=true .. here's a kinda messy comparison of all the data types between PG, DataFusion, and Arrow (with some/most of Spark and Parquet types filled in too). Clearly we should add support for casting interval. Other than that, I have no opinion currently.

**Comment by liukun4515** [comment]
> I fully understand and support that we are trying to mirror PG syntax. But I don't think that should prevent us from supporting a superset of data types that PG doesn't. Especially when the support already exists in Arrow and DataFusion functions and we just need this thin shim to extend the support to SQL.

Arrow and datafusion are two different project, we don't need to align function or feature for them.
I think datafusion is a sql/core lib, and datafusion can use other memory computing lib without the arrow-rs/arrow.

**Comment by liukun4515** [comment]
@kmitchener I really want to know what case you have to use the unsigned value

**Comment by alamb** [comment]
I plan to review / respond in more detail tomorrow. I won't have a chance
to do so today

**Comment by kmitchener** [comment]
> I plan to review / respond in more detail tomorrow. I won't have a chance to do so today

appreciate the feedback, thank you

**Comment by alamb** [comment]
I would summarize this issue as "sql engines like postgres and spark" avoid supporting signed and unsigned variants of types (presumably because for most usecases the difference in ranges are not important enough to justify the additional implementation and user comlexity).

However, Arrow **does** support unsigned variants (e.g `UInt64` and `Int64`, `UInt32` and `Int32`, etc) and DataFusion has a non trivial amount  of code to support those unsigned types (e.g. check out the uses of `UInt64`: https://github.com/apache/arrow-datafusion/search?q=UInt64)

I think it would be a reasonable discussion to start about 'should DataFusion support unsigned types at all' but given that datafusion does support those types internally, I think it is inconsistent to disallow them in the SQL syntax and I can't think of any reason not to support them

Thus after some additional thought I believe we should merge https://github.com/apache/arrow-datafusion/pull/3309

@liukun4515  if you disagree, do you have any additional concern you have (apart from unsigned types not being supported in spark or postgres)?

cc @liukun4515 @kmitchener 

I would also be interested in any other opinions from the community (@houqp  @andygrove )

**Comment by alamb** [comment]
I also sent a note in slack and to the dev@arrow.apache.org mailing list to solicit some additional opinions

**Comment by tustvold** [comment]
Does the DataFrame API currently support unsigned integers? I suspect it does. If so supporting them within SQL seems reasonable to me, but I'm no SQL expert.

My 2 cents is that the arrow specification likely is overly broad, who really needs 8 different types of dictionary encoding. However, its major selling point is portability, and only supporting a subset works against this. Whilst **I** may feel certain types are esoteric and unnecessary, they may be crucial to someone else's workload, and so we should endeavour to support what we reasonably can.

**Comment by avantgardnerio** [comment]
I'll weigh in here, since this appears to be an RFC. I see DataFusion as a query engine defined by its basis on Arrow. I generally think we should follow Postgres-flavored ANSI spec, however given the internals of the engine are part of its value proposition, I think it makes sense to extend where applicable to expose underlying functionality that would otherwise go to waste.

TLDR: +1 for supporting all arrow types.

**Comment by alamb** [comment]
> Does the DataFrame API currently support unsigned integers? I suspect it does. If so supporting them within SQL seems reasonable to me, but I'm no SQL expert.

The dataframe API itself is higher level and doesn't get into the level of DataTypes
https://docs.rs/datafusion/11.0.0/datafusion/dataframe/struct.DataFrame.html

The expressions API certainly does support all the arrow types (and thus unsigned). For example,  https://docs.rs/datafusion/11.0.0/datafusion/prelude/fn.cast.html would allow you to cast to an unsigned int like:

```rust
cast(input_expr, DataFusion::UInt64
```

**Comment by liukun4515** [comment]
> I would summarize this issue as "sql engines like postgres and spark" avoid supporting signed and unsigned variants of types (presumably because for most usecases the difference in ranges are not important enough to justify the additional implementation and user comlexity).
> 
> However, Arrow **does** support unsigned variants (e.g `UInt64` and `Int64`, `UInt32` and `Int32`, etc) and DataFusion has a non trivial amount of code to support those unsigned types (e.g. check out the uses of `UInt64`: https://github.com/apache/arrow-datafusion/search?q=UInt64)
> 
> I think it would be a reasonable discussion to start about 'should DataFusion support unsigned types at all' but given that datafusion does support those types internally, I think it is inconsistent to disallow them in the SQL syntax and I can't think of any reason not to support them
> 
> Thus after some additional thought I believe we should merge #3309
> 
> @liukun4515 if you disagree, do you have any additional concern you have (apart from unsigned types not being supported in spark or postgres)?
> 
> cc @liukun4515 @kmitchener
> 
> I would also be interested in any other opinions from the community (@houqp @andygrove )

I know in the core of datafusion we have supported some operation for the unsigned value.
It's good to support unsigned value for datafusion, but I just concern it will cause some undefined behavior.

If some users want this feature, I think it can be supported, and also need to update and add document to describe them.

@alamb @kmitchener

**Comment by liukun4515** [comment]
I will list some places that we need to handle the unsigned value:

- [ ] physical operation
- [ ] type coercion

**Comment by alamb** [comment]
Thanks @liukun4515  -- good idea. I made a tracking ticket for all the outstanding issues I know of for unsigned integer support in https://github.com/apache/arrow-datafusion/issues/3376 and added your items to it. 

It seems everyone who commented is in favor of supporting unsigned integers. Thank you all for your comments
